### PR TITLE
Combined PR

### LINKED ISSUE #15: Fill like a DropDownList
Is there a way to use this like a DropDownList?
Example when I click on it all list items are shown and then I can filter them.

### LINKED ISSUE #25: SearchMethod should be IEnumerable
Is there any reason that SearchMethod needs to be List<T> ? IEnumerable<T> is the best choice to collection parameters.

### LINKED ISSUE #26: onfocusout should hide the suggestions
Currently the suggestions are not hidden when the input loses focus. This won't be an easy fix if you only use C# though. Since the `onfocusout` event isn't fully supported in Blazor yet.

### LINKED ISSUE #27: After showing suggestions make the first item selectable by using tab
1. Search Item
2. Get suggestions
3. Use tab to select the first one.

### LINKED ISSUE #29: Endless onfocus loop
There is an endless loop occuring in the component.
To reproduce:
1. Set a breakpoint in the setter of the `_searchText` 
2. Optionally set the default `Debounce` value to `5000000` (so we know it's not coming from here.
=> endless onfocus loop

To fix this issue:
- change onfocus to onclick in the `input` of the control.

### PR DESCRIPTION
**Pull request by vertonghenb**
Basically this PR fixes multiple issues
- resolves #29
- resolves #27
- resolves #15
- resolves #25 
- resolves #26 
and combines following PR's:
- resolves #28
- resolves #30

### End Result:
- Notice the usage of `TAB`
- Comboboxstyle
- Loading is also supported
- Max items is set to 25 by default new parameter `MaximumSuggestions`
- Escape, OnFocusOut Supported
- Task<IEnumerable<T>> Instead of Task<List<T>>
typahead

### CODE REVIEW

**Comment by vertonghenb** [comment]
Ok, I'm gonna stop here, pretty happy with the result apart from:
- No Arrow Keys
- onfocusout does not close the searchlist

**Comment by vertonghenb** [comment]
Update:
- Escape supported
- Focusout Supported

**Comment by chrissainty** [comment]
@vertonghenb I'm seeing this error when using the server-side Sample. Have you noticed it?

Screenshot 2019-08-26 at 13 45 26

File: src/Blazored.Typeahead/BlazoredTypeahead.razor.cs [hunk 1]
```diff
@@ -11,26 +11,31 @@ namespace Blazored.Typeahead
...
         [Parameter] public RenderFragment FooterTemplate { get; set; }
         [Parameter] public int MinimumLength { get; set; } = 1;
         [Parameter] public int Debounce { get; set; } = 300;
+        [Parameter] public int MaximumSuggestions { get; set; } = 25;
 
         protected bool Searching { get; set; } = false;
-        protected bool EditMode { get; set; } = true;
-        protected List<TItem> SearchResults { get; set; } = new List<TItem>();
-
+        protected bool ShouldShowMenu { get; private set; } = false;
```
Comment (chrissainty) [inline src/Blazored.Typeahead/BlazoredTypeahead.razor.cs]: I'm not sure of the naming here, `Shouldxxx` doesn't seem to read well... `if (ShouldShowMenu) { // do stuff }`. What about a change to `ShowingMenu`... `if (ShowingMenu) { // do stuff }`. This is very much personal a preference thing.

File: src/Blazored.Typeahead/BlazoredTypeahead.razor.cs [hunk 2]
```diff
@@ -11,26 +11,31 @@ namespace Blazored.Typeahead
...
+        protected bool ShouldShowInput { get; private set; } = true;
```
Comment (chrissainty) [inline src/Blazored.Typeahead/BlazoredTypeahead.razor.cs]: Same as above

File: src/Blazored.Typeahead/BlazoredTypeahead.razor.cs [hunk 3]
```diff
@@ -11,26 +11,31 @@ namespace Blazored.Typeahead
...
+        protected bool ShouldShowMask { get; private set; } = false;
```
Comment (chrissainty) [inline src/Blazored.Typeahead/BlazoredTypeahead.razor.cs]: Same as above

File: src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs [hunk 4]
```diff
@@ -14,22 +14,28 @@ public class BlazoredTypeaheadInputBase<TItem> : InputBase<TItem>, IDisposable
...
         [Parameter] public RenderFragment FooterTemplate { get; set; }
         [Parameter] public int MinimumLength { get; set; } = 1;
         [Parameter] public int Debounce { get; set; } = 300;
+        [Parameter] public int MaximumSuggestions { get; set; } = 25;
 
         protected bool Searching { get; set; } = false;
-        protected bool EditMode { get; set; } = true;
-        protected List<TItem> SearchResults { get; set; } = new List<TItem>();
-
+        protected bool ShouldShowMenu { get; private set; } = false;
```
Comment (chrissainty) [inline src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs]: See comments on naming on the non-forms typeahead

File: src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs [hunk 5]
```diff
@@ -14,22 +14,28 @@ public class BlazoredTypeaheadInputBase<TItem> : InputBase<TItem>, IDisposable
...
+        protected bool ShouldShowInput { get; private set; } = true;
```
Comment (chrissainty) [inline src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs]: See comments on naming on the non-forms typeahead

File: src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs [hunk 6]
```diff
@@ -14,22 +14,28 @@ public class BlazoredTypeaheadInputBase<TItem> : InputBase<TItem>, IDisposable
...
+        protected bool ShouldShowMask { get; private set; } = false;
```
Comment (chrissainty) [inline src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs]: See comments on naming on the non-forms typeahead

File: src/Blazored.Typeahead/Interop.cs [hunk 7]
```diff
@@ -0,0 +1,39 @@
+﻿using Microsoft.AspNetCore.Components;
+using Microsoft.JSInterop;
+using System;
+using System.Threading.Tasks;
+
+namespace Blazored.Typeahead
+{
+    public static class Interop
```
Comment (chrissainty) [inline src/Blazored.Typeahead/Interop.cs]: I think this should be internal, I don't think we need to expose this out to consumers.

Comment (vertonghenb) [reply src/Blazored.Typeahead/Interop.cs]: A class that needs `JSInvokable` has to be static, else the JS won't be able to call in it I think.

File: src/Blazored.Typeahead/BlazoredTypeahead.razor.cs [hunk 8]
```diff
@@ -109,7 +109,7 @@ protected override async Task OnAfterRenderAsync()
                 await Interop.AddEscapeEventListener(JSRuntime, typeahead);
                 await Interop.AddFocusOutEventListener(JSRuntime, typeahead);
                 Interop.OnEscapeEvent += OnEscape;
-                Interop.OnEscapeEvent += OnFocusOut;
+                Interop.OnFocusOutEvent += OnFocusOut;
```
Comment (vertonghenb) [inline src/Blazored.Typeahead/BlazoredTypeahead.razor.cs]: This made the focusout event unavailable

File: src/Blazored.Typeahead/BlazoredTypeahead.razor.cs (hunk 1 above)
Comment (vertonghenb) [reply src/Blazored.Typeahead/BlazoredTypeahead.razor.cs]: Agreed, `IsShowing...` ?

**Comment by vertonghenb** [comment]
> @vertonghenb I'm seeing this error when using the server-side Sample. Have you noticed it?
> 
> Screenshot 2019-08-26 at 13 45 26

Yeah, I had a typo and it wasn't able to do anything. Can you pull the latest changes and check if it's fixed?
https://github.com/Blazored/Typeahead/pull/31#discussion_r317618589

File: src/Blazored.Typeahead/Interop.cs (hunk 7 above)
Comment (vertonghenb) [reply src/Blazored.Typeahead/Interop.cs]: Just tested it, has to be public.

File: src/Blazored.Typeahead/BlazoredTypeahead.razor.cs (hunk 8 above)
Comment (vertonghenb) [reply src/Blazored.Typeahead/BlazoredTypeahead.razor.cs]: FIxed.

File: src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs (hunk 6 above)
Comment (vertonghenb) [reply src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs]: protected bool IsSearching { get; private set; } = false;
        protected bool IsShowingSuggestions { get; private set; } = false;
        protected bool IsShowingSearchbar { get; private set; } = true;
        protected bool IsShowingMask { get; private set; } = false;

File: src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs (hunk 5 above)
Comment (vertonghenb) [reply src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs]: protected bool IsSearching { get; private set; } = false;
        protected bool IsShowingSuggestions { get; private set; } = false;
        protected bool IsShowingSearchbar { get; private set; } = true;
        protected bool IsShowingMask { get; private set; } = false;

File: src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs (hunk 4 above)
Comment (vertonghenb) [reply src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor.cs]: protected bool IsSearching { get; private set; } = false;
        protected bool IsShowingSuggestions { get; private set; } = false;
        protected bool IsShowingSearchbar { get; private set; } = true;
        protected bool IsShowingMask { get; private set; } = false;

File: src/Blazored.Typeahead/BlazoredTypeahead.razor.cs (hunk 2 above)
Comment (vertonghenb) [reply src/Blazored.Typeahead/BlazoredTypeahead.razor.cs]: protected bool IsSearching { get; private set; } = false;
        protected bool IsShowingSuggestions { get; private set; } = false;
        protected bool IsShowingSearchbar { get; private set; } = true;
        protected bool IsShowingMask { get; private set; } = false;

**Comment by vertonghenb** [comment]
Done any comments on the way the tabs/esc/delete etc works?

File: src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor [hunk 9]
```diff
@@ -1,8 +1,8 @@
 ﻿@typeparam TItem
 @inherits BlazoredTypeaheadInputBase<TItem>
 
-<div class="blazored-typeahead">
-    <div class="blazored-typeahead__controls" @ref="typeahead" @ref:suppressField>
```
Comment (vertonghenb) [inline src/Blazored.Typeahead/Forms/BlazoredTypeaheadInput.razor]: This was a problem on the Form Component, fixed now.

File: src/Blazored.Typeahead/Interop.cs (hunk 7 above)
Comment (chrissainty) [reply src/Blazored.Typeahead/Interop.cs]: Of course, it does!

**Comment by vertonghenb** [comment]
@chrissainty Sure, I'll create a new issue for the arrow keys, but not sure how to tackle that issue atm.
Note that this PR is a breaking change.

**Comment by vertonghenb** [comment]
Yup you can merge.

**Pull request merged**